Sources/Injection.swift

```diff
@@ -15,12 +16,12 @@ extension Dependency {
 final class DependencyContainer<T>: Dependency {
     let isSingleton: Bool
     let type: Any.Type
-    let factory: DependencyFactory
+    let factory: AnyDependencyFactory
 
-    init(isSingleton: Bool, factory: @escaping () -> T) {
+    init(isSingleton: Bool, factory: @escaping DependencyFactory<T>) {
```

Comment:
Perhaps this initializer could be marked with `rethrows` so that even the lazySingleton can rethrow, wdyt?

```suggestion
    init(isSingleton: Bool, factory: @escaping DependencyFactory<T>) rethrows {
```